src/librustc_mir/interpret/operand.rs

```diff
@@ -619,18 +619,23 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
                     Err(_) =>
                         throw_unsup!(InvalidDiscriminant(raw_discr.erase_tag())),
                 };
-                let real_discr = if discr_val.layout.ty.is_signed() {
+                let (discr_ty_signed, discr_ty_size) = match &rval.layout.ty.sty {
+                    ty::Adt(adt, _) => {
```

Comment:
This seems fishy. In my experience it is basically never correct to look at the `Ty` of something in Miri, with very few exceptions.

Reply:
OTOH, this looks like you basically inlined old `discr_type`? Why can't we use that any more / extend that to also return signedness?

Reply:
What old `discr_type`? I hoisted the `discr_type` call, and note that it only made sense for ADTs, whereas generators can also have discriminants (but they're not signed, so they wouldn't hit the signed logic previously).

Reply:
Oh, `discr_type` only works for ADTs, of course.

Maybe this amtch here should move to `Ty`, so that we can just do `layout.ty.discr_type()`? Its return type `IntType` (yet another one, next to `layout::Integer`, `IntTy` and `UintTy`) seems to have just the right amount of freedom.

Reply:
I think we already have something for variant names, so this seems... okay-ish?
Maybe the generator case can just be a `_ =>` so we really only special-case ADTs with signed discriminants.

I don't know, I really don't like signed integers :(.